internal/services/notification/commitstatusdelivery.go

```diff
@@ -0,0 +1,235 @@
+	l := n.lf.NewLock(CommitStatusDeliveriesLockKey)
+	if err := l.TryLock(ctx); err != nil {
+		if errors.Is(err, lock.ErrLocked) {
+			return nil
+		}
+		return errors.WithStack(err)
+	}
+	defer func() { _ = l.Unlock() }()
+
+	curCommitStatusDeliverySequence := uint64(0)
+
+	for {
+		var commitStatusDeliveries []*types.CommitStatusDelivery
+
+		err := n.d.Do(ctx, func(tx *sql.Tx) error {
+			var err error
+			commitStatusDeliveries, err = n.d.GetCommitStatusDeliveriesAfterSequence(tx, curCommitStatusDeliverySequence, types.DeliveryStatusNotDelivered, MaxCommitStatusDeliveriesQueryLimit)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+
+			return nil
+		})
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		for _, c := range commitStatusDeliveries {
+			if err := n.handleCommitStatusDelivery(ctx, c.ID); err != nil {
+				n.log.Err(err).Msgf("failed to trigger commit status data delivery event")
+			}
+
+			curCommitStatusDeliverySequence = c.Sequence
+		}
+
+		if len(commitStatusDeliveries) < MaxCommitStatusDeliveriesQueryLimit {
+			return nil
+		}
+	}
+}
+
+func (n *NotificationService) handleCommitStatusDelivery(ctx context.Context, commitStatusDeliveryID string) error {
+	var commitStatus *types.CommitStatus
+
+	err := n.d.Do(ctx, func(tx *sql.Tx) error {
+		var err error
+
+		commitStatusDelivery, err := n.d.GetCommitStatusDeliveryByID(tx, commitStatusDeliveryID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		if commitStatusDelivery.DeliveryStatus != types.DeliveryStatusNotDelivered {
+			return nil
+		}
+
+		commitStatus, err = n.d.GetCommitStatusByID(tx, commitStatusDelivery.CommitStatusID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if commitStatus == nil {
+		return nil
+	}
+
+	delivered, err := n.u.updateCommitStatus(ctx, commitStatus)
+	// err != nil is considered a failed delivery
+	if err != nil {
+		n.log.Err(err).Send()
+	}
+
+	var deliveredAt *time.Time
+	if delivered {
+		deliveredAt = util.TimeP(time.Now())
+	}
+
+	err = n.d.Do(ctx, func(tx *sql.Tx) error {
+		var err error
+
+		commitStatusDelivery, err := n.d.GetCommitStatusDeliveryByID(tx, commitStatusDeliveryID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		if commitStatusDelivery.DeliveryStatus != types.DeliveryStatusNotDelivered {
+			return nil
+		}
+
+		if delivered {
+			commitStatusDelivery.DeliveryStatus = types.DeliveryStatusDelivered
+		} else {
+			commitStatusDelivery.DeliveryStatus = types.DeliveryStatusDeliveryError
+		}
+
+		commitStatusDelivery.DeliveredAt = deliveredAt
+
+		if err = n.d.UpdateCommitStatusDelivery(tx, commitStatusDelivery); err != nil {
+			return errors.WithStack(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+type GitSourceCommitStatusUpdater struct {
+	n *NotificationService
```

Comment:
Don't set notification service inside status updater. Set the configstore client, config etc....